akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ExternalInteractions.scala

```diff
@@ -151,16 +152,17 @@ private[akka] trait SnapshotInteractions[C, E, S] {
         setup.selfUntyped)
   }
 
-  /** Deletes the snapshot identified by `sequenceNr`. */
+  /** Deletes the snapshots up to and including the `sequenceNr`. */
   protected def internalDeleteSnapshots(toSequenceNr: Long): Unit = {
-    setup.log.debug("Deleting snapshot  to [{}]", toSequenceNr)
-
-    val deleteTo = toSequenceNr - 1
-    val deleteFrom = math.max(0, setup.retention.toSequenceNumber(deleteTo))
-    val snapshotCriteria = SnapshotSelectionCriteria(minSequenceNr = deleteFrom, maxSequenceNr = deleteTo)
-
-    setup.log.debug("Deleting snapshots from [{}] to [{}]", deleteFrom, deleteTo)
-    setup.snapshotStore
-      .tell(SnapshotProtocol.DeleteSnapshots(setup.persistenceId.id, snapshotCriteria), setup.selfUntyped)
+    if (toSequenceNr > 0) {
+      // We could use 0 as fromSequenceNr to delete all older snapshots, but that might be inefficient for
+      // large ranges depending on how it's implemented in the snapshot plugin. Therefore we use the
+      // same window as defined for how much to keep in the retention criteria
+      val fromSequenceNr = setup.retention.toSequenceNumber(toSequenceNr)
+      val snapshotCriteria = SnapshotSelectionCriteria(minSequenceNr = fromSequenceNr, maxSequenceNr = toSequenceNr)
+      setup.log.debug("Deleting snapshots from sequenceNr [{}] to [{}]", fromSequenceNr, toSequenceNr)
+      setup.snapshotStore
+        .tell(SnapshotProtocol.DeleteSnapshots(setup.persistenceId.id, snapshotCriteria), setup.selfUntyped)
+    }
```

Comment:
Not sure how I missed this before but this confuses me: we call `Retention.toSequenceNumber(toSequenceNr)` and catch that in `fromSequenceNr`, so the `toSequenceNumber` method is in fact giving us a safe `from`? Can naming make this more clear?


Also, would it be a bug if we ever passed 0 as `toSequenceNr` here? Should we fail rather than silently swallow it?

Reply:
> fromSequenceNr

I tried to explain in the comment above. It's just a way to create a sliding window.
I agree that the naming of that `Retention.toSequenceNumber` is confusing since it's used both for finding the real `to`, but also to find the `from`. I'll add that to ticket https://github.com/akka/akka/issues/26545#issuecomment-475629827

> Also, would it be a bug if we ever passed 0 as toSequenceNr here? Should we fail rather than silently swallow it?